AdvancedTaskPart2SpecFlowProject/StepDefinitions/CertificationStepDefinitions.cs

```diff
@@ -0,0 +1,172 @@
+            expectedMessage = "This information is already exist.";
+        }
+
+        [When(@"user adds an education record with already existing data in Text Boxes and selecting different year from dropdown")]
+        public void WhenUserAddsAnEducationRecordWithAlreadyExistingDataInTextBoxesAndSelectingDifferentYearFromDropdown()
+        {
+            certificationAddAndDeleteComponent.DeleteAllCertificationRecords();
+            certificationAddAndDeleteComponent.AddCertification(certificationDMList[1]);
+            certificationAddAndDeleteComponent.AddCertification(certificationDMList[5]);
+            actualMessage = certificationRenderComponents!.CapturePopupMessage();
+            expectedMessage = "Duplicated data";
+        }
+
+        [When(@"user adds an certification record with new data in Text Boxes and selecting already existing year from dropdown")]
+        public void WhenUserAddsAnCertificationRecordWithNewDataInTextBoxesAndSelectingAlreadyExistingYearFromDropdown()
+        {
+            certificationAddAndDeleteComponent.DeleteAllCertificationRecords();
+            certificationAddAndDeleteComponent.AddCertification(certificationDMList[1]);
+            certificationAddAndDeleteComponent.AddCertification(certificationDMList[6]);
+            actualMessage = certificationRenderComponents!.CapturePopupMessage();
+            expectedMessage = certificationDMList[6].certificateName + " has been added to your certification";
+        }
+
+        [When(@"user adds a certification record with Special Characters and numbers in certification name TextBox")]
+        public void WhenUserAddsACertificationRecordWithSpecialCharactersAndNumbersInCertificationNameTextBox()
+        {
+            certificationAddAndDeleteComponent.DeleteAllCertificationRecords();
+            certificationAddAndDeleteComponent.AddCertification(certificationDMList[7]);
+            actualMessage = certificationRenderComponents!.CapturePopupMessage();
+            expectedMessage = certificationDMList[7].certificateName + " has been added to your certification";
+        }
+
+        [When(@"user adds a certification record with very long data in CertifiedFrom TextBox")]
+        public void WhenUserAddsACertificationRecordWithVeryLongDataInCertifiedFromTextBox()
+        {
+            certificationAddAndDeleteComponent.DeleteAllCertificationRecords();
+            certificationAddAndDeleteComponent.AddCertification(certificationDMList[8]);
+            actualMessage = certificationRenderComponents!.CapturePopupMessage();
+            expectedMessage = certificationDMList[8].certificateName + " has been added to your certification";
+        }
+
+        [When(@"user adds a certification record with only Spaces in TextBoxes")]
+        public void WhenUserAddsACertificationRecordWithOnlySpacesInTextBoxes()
+        {
+            certificationAddAndDeleteComponent.DeleteAllCertificationRecords();
```

Comment:
so i see that you thinking about state but you could take this one set forward!

example before adding language you need to make sure there a no existing languages as this could result in false positive test pass
you will need to think about this for all the scenarios

the state is say that you have a test that is adding a language english, the second time you run this test you should delete any existing language so that the state is reset, this will guarantee that your test does not produce false negatives and passes always
you can assume that test execution order is not always guaranteed
imagine if you run the tests  in parallel mode the all tests will run at the same time
so you need to also make sure you do not wipe out other tests data
hence why when you remove data the best thing to do is delete only data you added for the test to be correct
e.g test 1 added english and french as language test 2 added hindi then when you reset data you already know what language you are adding for each test
there for on reset of the two test you should delete english and french for test 1 and hindi for test 2

So to sum up you should delete all data if there is any before the start of any test and then after each test you should delete the data that a particular test is adding!
This is ensure that upon staring the test pack you have reset state and upon finishing each test deletes their data that they added!

Reply:
Hi Aman,

I am resetting the state by deleting all the data at the start of every test. Then why should we clear the state at the end of each test? No tests are dependent on each other or the test order does not affect the test result as each test starts from state 0. If we clear the state at the end of each test, then every screenshot will be same as every test ends with a delete all.